### ISSUE TITLE
Changes in PR #18665 will break exported native modules

### ISSUE THREAD

**Issue by tn0502** [DefinitelyTyped/DefinitelyTyped#18734]
- [x] I tried using the `@types/react-native` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @alloy @gyzerok @huhuanming @iRoachie @timwangdev @kamal

A recently merged PR #18665 will break for us since exported native modules rely on `NativeModules` to be of type `any` otherwise the compiler will complain about missing properties:

```
error TS2339: Property 'MyNativeModule' does not exist on type 'NativeModulesStatic'.
error TS2339: Property 'SettingsManager' does not exist on type 'NativeModulesStatic'.
error TS2339: Property 'I18nManager' does not exist on type 'NativeModulesStatic'.
```

**Comment by tn0502** [comment]
cc @SahinVardar

**Comment by huhuanming** [comment]
```
declare interface INativeModules {
  Tracker: INativeModlueTracker
}
```

```
import { NativeModules } from 'react-native'

const Tracker = (NativeModules as INativeModules).Tracker
```

**Comment by tn0502** [comment]
Are you saying you want users to add a declaration for every native module they have, and then do a type cast everywhere they're used?

**Comment by huhuanming** [comment]
Has no good choice and the better way is not found now.

**Comment by huhuanming** [comment]
TypeScript don't allow adding new method or interface into a exist class.

Maybe we need a feature like `Category` in `Objective-C`, or 'MixIn' in other language.

**Comment by huhuanming** [comment]
The other good practice now is wrapping the `NativeModules`, and exposes a custom `NativeModules`.
like this.

in `CustomNativeModules.ts`.
```
import { NativeModules } from 'react-native'

const CustomNativeModules: INativeModules =   NativeModules as INativeModules

export default CustomNativeModules
```

and using CustomNativeModules in other files.

**Comment by SahinVardar** [comment]
@tn0502 
A quick fix in your case would be 

```ts
import { NativeModules as any } from 'react-native'
```

This way, you would have your previous condition and no type information at all for your NativeModules. 

A better approach would be to create typings for your native modules. This could look like this in a type script definition file. 

```ts
export * from 'react-native'

declare module 'react-native' {
    export interface NativeModulesStatic {
        MyNativeModule: MyNativeModuleType;
        SettingsManager: SettingsManagerType;
        I18nManager: I18nManagerType;
    }
}
```

Then you would have a typing for your native modules. In my opinion, since we are working in type script anyways, we should have types for almost everything, otherwise I don't see the benefit of type script. 

I can understand that forcing people either to do typings for their native modules or casting NativeModules to any is not super nice, but losing the ability to do typing for NativeModules would be also a huge lost which I would like to avoid.

**Comment by huhuanming** [comment]
thx, but not working for me.

**Comment by SahinVardar** [comment]
@huhuanming 
Can you provide some details ?

**Comment by abrahambotros** [comment]
@SahinVardar you can't do the following (which is why I believe @huhuanming said it wasn't working):

```
import { NativeModules as any } from 'react-native'
```

That renames `NativeModules` with the alias `any`. Not sure if TS will allow you to overwrite this in the first place, but even if so, you definitely don't want `NativeModules` actually aliased to `any`.

Maybe I'm missing something, but what's a concrete example where we actually want to typed to `{}` exactly, and not `any` as before? Seems like if you want stricter typing for your `NativeModules` import, you'd just do what @huhuanming suggested at https://github.com/DefinitelyTyped/DefinitelyTyped/issues/18734#issuecomment-320888823.

Lastly, https://github.com/DefinitelyTyped/DefinitelyTyped/pull/18665 is mainly for `react-native-sensor-manager`, and not `react-native` - best practice would be not to make changes to `react-native` as a side note of a different package in the future.

My vote is to revert this specific change and make a new release ASAP with `NativeModules` typed as `any` as it was before. Thoughts @tn0502?

**Comment by SahinVardar** [comment]
```
import { NativeModules as any } from 'react-native'
```

Yeah this was totally not what I though it was ... totally wrong.

@abrahambotros How about [this],
(https://github.com/DefinitelyTyped/DefinitelyTyped/pull/18665#commitcomment-23606379) solution mentioned by @newyankeecodeshop ? I would prefer that, so we can type native modules but do not force anyone to do so.

**Comment by SahinVardar** [comment]
Create a pull request to fix the issue. See [here](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/19435#issuecomment-325756557)

**Comment by alloy** [comment]
Closed by #19435.
